File: Packs/Obsidian/src/scripts/vault_ops.py
# /// script
# dependencies = []
# ///
"""Advanced vault operations — bulk property edits, link analysis, structure validation.

Usage:
  vault_ops.py property-set PROPERTY VALUE [--folder FOLDER] [--filter PROP=VAL]
  vault_ops.py property-remove PROPERTY [--folder FOLDER]
  vault_ops.py property-stats [--folder FOLDER]
  vault_ops.py link-graph [--folder FOLDER] [--format dot|json]
  vault_ops.py structure-check [--expected FOLDER1,FOLDER2,...]
  vault_ops.py frontmatter-fix [--folder FOLDER] [--dry-run]
"""

import argparse
import json
import os
import re
import sys
from collections import Counter, defaultdict
from pathlib import Path


# -- Config -------------------------------------------------------------------

def load_env() -> None:
    env_path = Path.home() / ".env"
    if not env_path.exists():
        return
    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, _, value = line.partition("=")
            os.environ.setdefault(key.strip(), value.strip())


load_env()
VAULT_ROOT = Path(os.environ.get("OBSIDIAN_VAULT_PATH", ""))
IGNORE_DIRS = {".obsidian", ".git", ".archive", ".trash", ".claude", ".vscode", ".docs", "node_modules"}


def _iter_notes(folder: str | None = None) -> list[Path]:
    """Iterate markdown files in vault, optionally filtered by folder."""
    root = VAULT_ROOT / folder if folder else VAULT_ROOT
    notes = []
    for md in sorted(root.rglob("*.md")):
        if any(p in IGNORE_DIRS for p in md.relative_to(VAULT_ROOT).parts):
            continue
        notes.append(md)
    return notes


def _parse_frontmatter_raw(path: Path) -> tuple[str | None, str | None, str | None]:
    """Return (pre_fm, fm_content, post_fm) or (None, None, full_content)."""
    text = path.read_text(encoding="utf-8", errors="ignore")
    if not text.startswith("---"):
        return None, None, text
    end = text.find("\n---", 3)
    if end == -1:
        return None, None, text
    pre = "---\n"
    fm = text[4:end]
    post = text[end + 4:]
    return pre, fm, post


def _write_frontmatter(path: Path, pre: str, fm: str, post: str) -> None:
    path.write_text(pre + fm + "\n---" + post, encoding="utf-8")


# -- Commands -----------------------------------------------------------------

def cmd_property_set(args) -> None:
    """Set a property on all notes in a folder."""
    notes = _iter_notes(args.folder)
    prop = args.property
    value = args.value
    updated = 0
    skipped = 0

    for note in notes:
        pre, fm, post = _parse_frontmatter_raw(note)
        if fm is None:
            # No frontmatter — add it
            path_content = post or ""
            note.write_text(f"---\n{prop}: {value}\n---\n{path_content}", encoding="utf-8")
            updated += 1
            continue

        # Check filter
        if args.filter:
            fkey, _, fval = args.filter.partition("=")
            pattern = re.compile(rf'^{re.escape(fkey.strip())}:\s*{re.escape(fval.strip())}', re.MULTILINE)
            if not pattern.search(fm):
                skipped += 1
                continue

        # Check if property exists
        prop_pattern = re.compile(rf'^{re.escape(prop)}:.*$', re.MULTILINE)
        if prop_pattern.search(fm):
            fm = prop_pattern.sub(f"{prop}: {value}", fm)
        else:
            fm = fm.rstrip("\n") + f"\n{prop}: {value}"

        _write_frontmatter(note, pre, fm, post)
        updated += 1

    print(f"Updated: {updated}, Skipped: {skipped}")


def cmd_property_remove(args) -> None:
    """Remove a property from all notes in a folder."""
    notes = _iter_notes(args.folder)
    prop = args.property
    removed = 0

    for note in notes:
        pre, fm, post = _parse_frontmatter_raw(note)
        if fm is None:
            continue

        prop_pattern = re.compile(rf'^{re.escape(prop)}:.*\n?', re.MULTILINE)
        new_fm = prop_pattern.sub("", fm)
        if new_fm != fm:
            _write_frontmatter(note, pre, new_fm, post)
            removed += 1

    print(f"Removed '{prop}' from {removed} notes")


def cmd_property_stats(args) -> None:
    """Show property usage statistics."""
    notes = _iter_notes(args.folder)
    prop_counts: Counter = Counter()
    prop_values: dict[str, Counter] = defaultdict(Counter)
    total = 0

    for note in notes:
        pre, fm, post = _parse_frontmatter_raw(note)
        if fm is None:
            continue
        total += 1
        for line in fm.split("\n"):
            if ":" in line and not line.startswith(" ") and not line.startswith("-"):
                key, _, val = line.partition(":")
                key = key.strip()
                val = val.strip().strip('"').strip("'")
                prop_counts[key] += 1
                if val and val != "[]":
                    prop_values[key][val] += 1

    print(f"\nProperty stats ({total} notes in {args.folder or 'vault'}):")
    print(f"{'Property':<25} {'Count':>6} {'Coverage':>8}  Top values")
    print("-" * 80)
    for prop, count in prop_counts.most_common(20):
        coverage = f"{count/total*100:.0f}%" if total else "0%"
        top = prop_values[prop].most_common(3)
        top_str = ", ".join(f"{v}({c})" for v, c in top) if top else "(empty)"
        print(f"  {prop:<23} {count:>6} {coverage:>8}  {top_str[:50]}")


def cmd_link_graph(args) -> None:
    """Generate a link graph in DOT or JSON format."""
    notes = _iter_notes(args.folder)
    stem_to_path: dict[str, str] = {}
    edges: list[tuple[str, str]] = []

    for note in notes:
        rel = str(note.relative_to(VAULT_ROOT))
        stem_to_path[note.stem] = rel

    for note in notes:
        rel = str(note.relative_to(VAULT_ROOT))
        try:
            text = note.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue
        for match in re.finditer(r"\[\[([^\]|]+)(?:\|[^\]]+)?\]\]", text):
            target = match.group(1).strip()
            if target in stem_to_path:
                edges.append((rel, stem_to_path[target]))

    if args.format == "dot":
        print("digraph vault {")
        print("  rankdir=LR;")
        seen_nodes = set()
        for src, dst in edges:
            for n in (src, dst):
                if n not in seen_nodes:
                    label = Path(n).stem
                    print(f'  "{n}" [label="{label}"];')
                    seen_nodes.add(n)
            print(f'  "{src}" -> "{dst}";')
        print("}")
    else:
        graph = {"nodes": list(stem_to_path.values()), "edges": [{"from": s, "to": d} for s, d in edges]}
        print(json.dumps(graph, indent=2))

    print(f"\n# {len(stem_to_path)} nodes, {len(edges)} edges", file=sys.stderr)


def cmd_structure_check(args) -> None:
    """Validate vault structure against conventions."""
    vault = VAULT_ROOT
    issues: list[str] = []

    # Expected top-level folders (configurable via --expected flag)
    if args.expected:
        expected = set(args.expected.split(","))
    else:
        # Auto-detect: just check for loose files, empty dirs, nested .obsidian
        expected = set()

    actual = {d.name for d in vault.iterdir() if d.is_dir() and not d.name.startswith(".")}

    if expected:
        missing = expected - actual
        extra = actual - expected
        if missing:
            issues.append(f"Missing expected folders: {', '.join(sorted(missing))}")
        if extra:
            issues.append(f"Extra top-level folders: {', '.join(sorted(extra))}")

    # Root-level loose files
    root_files = [f.name for f in vault.iterdir() if f.is_file() and not f.name.startswith(".")]
    if root_files:
        issues.append(f"Loose files at root: {', '.join(root_files)}")

    # Check for notes without frontmatter in common content folders
    for folder_name in ["Areas", "Projects"]:
        folder = vault / folder_name
        if folder.exists():
            no_fm = 0
            for md in folder.glob("*.md"):
                text = md.read_text(encoding="utf-8", errors="ignore")
                if not text.startswith("---"):
                    no_fm += 1
            if no_fm:
                issues.append(f"{folder_name}: {no_fm} notes without frontmatter")

    # Empty folders
    for d in vault.rglob("*"):
        if not d.is_dir():
            continue
        if any(p in IGNORE_DIRS for p in d.relative_to(vault).parts):
            continue
        if not any(d.iterdir()):
            issues.append(f"Empty folder: {d.relative_to(vault)}")

    # Nested .obsidian
    for obs in vault.rglob(".obsidian"):
        if obs.parent != vault:
            issues.append(f"Nested .obsidian config: {obs.relative_to(vault)}")

    if issues:
        print(f"Found {len(issues)} issues:")
        for issue in issues:
            print(f"  - {issue}")
    else:
        print("Vault structure OK")


def cmd_frontmatter_fix(args) -> None:
    """Find and report frontmatter issues (unclosed, missing, malformed)."""
    notes = _iter_notes(args.folder)
    issues = []

    for note in notes:
        rel = note.relative_to(VAULT_ROOT)
        text = note.read_text(encoding="utf-8", errors="ignore")

        if not text.startswith("---"):
            if text.strip().startswith("---"):
                issues.append((rel, "frontmatter preceded by whitespace"))
            elif "---" in text[:200]:
                issues.append((rel, "frontmatter not on line 1"))
            else:
                issues.append((rel, "no frontmatter"))
            continue

        end = text.find("\n---", 3)
        if end == -1:
            issues.append((rel, "unclosed frontmatter (no closing ---)"))
            if not args.dry_run:
                # Find where YAML ends
                lines = text.split("\n")
                yaml_end = 1
                for i in range(1, len(lines)):
                    line = lines[i]
                    if re.match(r'^(\w[\w\s-]*:.*|\s+-\s+.*|\s+.*|\s*)$', line):
                        yaml_end = i + 1
                    else:
                        break
                lines.insert(yaml_end, "---")
                note.write_text("\n".join(lines), encoding="utf-8")
                issues[-1] = (rel, "unclosed frontmatter — FIXED")

    if issues:
        print(f"Found {len(issues)} frontmatter issues:")
        for rel, issue in issues:
            print(f"  {rel}: {issue}")
    else:
        print("All frontmatter OK")


# -- Entry point --------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(description="Advanced Obsidian vault operations")
    sub = parser.add_subparsers(dest="command", required=True)

    p_set = sub.add_parser("property-set", help="Set a property on notes")
    p_set.add_argument("property")
    p_set.add_argument("value")
    p_set.add_argument("--folder", default=None)
    p_set.add_argument("--filter", default=None, help="Only update notes matching PROP=VAL")

    p_rm = sub.add_parser("property-remove", help="Remove a property from notes")
    p_rm.add_argument("property")
    p_rm.add_argument("--folder", default=None)

    p_stats = sub.add_parser("property-stats", help="Show property usage stats")
    p_stats.add_argument("--folder", default=None)

    p_graph = sub.add_parser("link-graph", help="Generate link graph")
    p_graph.add_argument("--folder", default=None)
    p_graph.add_argument("--format", default="json", choices=["dot", "json"])

    p_check = sub.add_parser("structure-check", help="Validate vault structure")
    p_check.add_argument("--expected", default=None, help="Comma-separated expected top-level folders")

    p_fix = sub.add_parser("frontmatter-fix", help="Find/fix frontmatter issues")
    p_fix.add_argument("--folder", default=None)
    p_fix.add_argument("--dry-run", action="store_true")

    args = parser.parse_args()
    {
        "property-set": cmd_property_set,
        "property-remove": cmd_property_remove,
        "property-stats": cmd_property_stats,
        "link-graph": cmd_link_graph,
        "structure-check": cmd_structure_check,
        "frontmatter-fix": cmd_frontmatter_fix,
    }[args.command](args)


if __name__ == "__main__":
    main()

File: Packs/Obsidian/src/scripts/vault_ops.ts
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

type PropertySetArgs = { command: "property-set"; property: string; value: string; folder: string | null; filter: string | null };
type PropertyRemoveArgs = { command: "property-remove"; property: string; folder: string | null };
type PropertyStatsArgs = { command: "property-stats"; folder: string | null };
type LinkGraphArgs = { command: "link-graph"; folder: string | null; format: "dot" | "json" };
type StructureCheckArgs = { command: "structure-check"; expected: string | null };
type FrontmatterFixArgs = { command: "frontmatter-fix"; folder: string | null; dryRun: boolean };
type Args = PropertySetArgs | PropertyRemoveArgs | PropertyStatsArgs | LinkGraphArgs | StructureCheckArgs | FrontmatterFixArgs;

const IGNORE_DIRS = new Set([".obsidian", ".git", ".archive", ".trash", ".claude", ".vscode", ".docs", "node_modules"]);

function loadEnv(): void {
  const envPath = path.join(os.homedir(), ".env");
  if (!fs.existsSync(envPath)) return;
  const content = fs.readFileSync(envPath, "utf8");
  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const idx = line.indexOf("=");
    const key = line.slice(0, idx).trim();
    const value = line.slice(idx + 1).trim();
    if (process.env[key] === undefined) process.env[key] = value;
  }
}

function decodeUtf8Ignore(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length;) {
    const b1 = bytes[i]!;
    if (b1 < 0x80) {
      out += String.fromCharCode(b1);
      i += 1;
    } else if (b1 >= 0xc2 && b1 <= 0xdf && i + 1 < bytes.length && (bytes[i + 1]! & 0xc0) === 0x80) {
      out += String.fromCodePoint(((b1 & 0x1f) << 6) | (bytes[i + 1]! & 0x3f));
      i += 2;
    } else if (b1 >= 0xe0 && b1 <= 0xef && i + 2 < bytes.length) {
      const b2 = bytes[i + 1]!;
      const b3 = bytes[i + 2]!;
      const valid = (b2 & 0xc0) === 0x80 && (b3 & 0xc0) === 0x80
        && !(b1 === 0xe0 && b2 < 0xa0)
        && !(b1 === 0xed && b2 >= 0xa0);
      if (valid) {
        out += String.fromCodePoint(((b1 & 0x0f) << 12) | ((b2 & 0x3f) << 6) | (b3 & 0x3f));
        i += 3;
      } else {
        i += 1;
      }
    } else if (b1 >= 0xf0 && b1 <= 0xf4 && i + 3 < bytes.length) {
      const b2 = bytes[i + 1]!;
      const b3 = bytes[i + 2]!;
      const b4 = bytes[i + 3]!;
      const valid = (b2 & 0xc0) === 0x80 && (b3 & 0xc0) === 0x80 && (b4 & 0xc0) === 0x80
        && !(b1 === 0xf0 && b2 < 0x90)
        && !(b1 === 0xf4 && b2 >= 0x90);
      if (valid) {
        out += String.fromCodePoint(((b1 & 0x07) << 18) | ((b2 & 0x3f) << 12) | ((b3 & 0x3f) << 6) | (b4 & 0x3f));
        i += 4;
      } else {
        i += 1;
      }
    } else {
      i += 1;
    }
  }
  return out;
}

function readTextIgnore(filePath: string): string {
  return decodeUtf8Ignore(fs.readFileSync(filePath));
}

loadEnv();

const VAULT_ROOT = process.env.OBSIDIAN_VAULT_PATH ?? "";

function relPath(full: string): string {
  return path.relative(VAULT_ROOT, full).split(path.sep).join("/");
}

function iterNotes(folder: string | null = null): string[] {
  const root = folder ? path.join(VAULT_ROOT, folder) : VAULT_ROOT;
  const notes: string[] = [];
  function visit(dir: string): void {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        visit(full);
      } else if (entry.isFile() && entry.name.endsWith(".md")) {
        const parts = relPath(full).split("/");
        if (parts.some((part) => IGNORE_DIRS.has(part))) continue;
        notes.push(full);
      }
    }
  }
  visit(root);
  return notes.sort();
}

function parseFrontmatterRaw(filePath: string): [string | null, string | null, string | null] {
  const text = readTextIgnore(filePath);
  if (!text.startsWith("---")) return [null, null, text];
  const end = text.indexOf("\n---", 3);
  if (end === -1) return [null, null, text];
  const pre = "---\n";
  const fm = text.slice(4, end);
  const post = text.slice(end + 4);
  return [pre, fm, post];
}

function writeFrontmatter(filePath: string, pre: string, fm: string, post: string): void {
  fs.writeFileSync(filePath, pre + fm + "\n---" + post, "utf8");
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function cmdPropertySet(args: PropertySetArgs): void {
  const notes = iterNotes(args.folder);
  const prop = args.property;
  const value = args.value;
  let updated = 0;
  let skipped = 0;

  for (const note of notes) {
    let [pre, fm, post] = parseFrontmatterRaw(note);
    if (fm === null) {
      const pathContent = post ?? "";
      fs.writeFileSync(note, `---\n${prop}: ${value}\n---\n${pathContent}`, "utf8");
      updated += 1;
      continue;
    }

    if (args.filter) {
      const idx = args.filter.indexOf("=");
      const fkey = idx === -1 ? args.filter : args.filter.slice(0, idx);
      const fval = idx === -1 ? "" : args.filter.slice(idx + 1);
      const pattern = new RegExp(`^${escapeRegExp(fkey.trim())}:\\s*${escapeRegExp(fval.trim())}`, "m");
      if (!pattern.test(fm)) {
        skipped += 1;
        continue;
      }
    }

    const propPattern = new RegExp(`^${escapeRegExp(prop)}:.*$`, "m");
    if (propPattern.test(fm)) {
      fm = fm.replace(propPattern, `${prop}: ${value}`);
    } else {
      fm = fm.replace(/\n+$/g, "") + `\n${prop}: ${value}`;
    }
    writeFrontmatter(note, pre!, fm, post!);
    updated += 1;
  }

  console.log(`Updated: ${updated}, Skipped: ${skipped}`);
}

function cmdPropertyRemove(args: PropertyRemoveArgs): void {
  const notes = iterNotes(args.folder);
  const prop = args.property;
  let removed = 0;
  for (const note of notes) {
    const [pre, fm, post] = parseFrontmatterRaw(note);
    if (fm === null) continue;
    const propPattern = new RegExp(`^${escapeRegExp(prop)}:.*\\n?`, "gm");
    const newFm = fm.replace(propPattern, "");
    if (newFm !== fm) {
      writeFrontmatter(note, pre!, newFm, post!);
      removed += 1;
    }
  }
  console.log(`Removed '${prop}' from ${removed} notes`);
}

function increment(map: Map<string, number>, key: string): void {
  map.set(key, (map.get(key) ?? 0) + 1);
}

function mostCommon(map: Map<string, number>, limit: number): Array<[string, number]> {
  return [...map.entries()].sort((a, b) => b[1] - a[1]).slice(0, limit);
}

function cmdPropertyStats(args: PropertyStatsArgs): void {
  const notes = iterNotes(args.folder);
  const propCounts = new Map<string, number>();
  const propValues = new Map<string, Map<string, number>>();
  let total = 0;

  for (const note of notes) {
    const [_pre, fm, _post] = parseFrontmatterRaw(note);
    if (fm === null) continue;
    total += 1;
    for (const line of fm.split("\n")) {
      if (line.includes(":") && !line.startsWith(" ") && !line.startsWith("-")) {
        const idx = line.indexOf(":");
        const key = line.slice(0, idx).trim();
        let val = line.slice(idx + 1).trim();
        val = stripChars(stripChars(val, "\""), "'");
        increment(propCounts, key);
        if (val && val !== "[]") {
          const values = propValues.get(key) ?? new Map<string, number>();
          increment(values, val);
          propValues.set(key, values);
        }
      }
    }
  }

  console.log(`\nProperty stats (${total} notes in ${args.folder ?? "vault"}):`);
  console.log(`${"Property".padEnd(25, " ")} ${"Count".padStart(6, " ")} ${"Coverage".padStart(8, " ")}  Top values`);
  console.log("-".repeat(80));
  for (const [prop, count] of mostCommon(propCounts, 20)) {
    const coverage = total ? `${(count / total * 100).toFixed(0)}%` : "0%";
    const values = propValues.get(prop);
    const top = values ? mostCommon(values, 3) : [];
    const topStr = top.length ? top.map(([value, valueCount]) => `${value}(${valueCount})`).join(", ") : "(empty)";
    console.log(`  ${prop.padEnd(23, " ")} ${String(count).padStart(6, " ")} ${coverage.padStart(8, " ")}  ${topStr.slice(0, 50)}`);
  }
}

function stripChars(value: string, char: string): string {
  let start = 0;
  let end = value.length;
  while (start < end && value[start] === char) start += 1;
  while (end > start && value[end - 1] === char) end -= 1;
  return value.slice(start, end);
}

function asciiJson(value: unknown): string {
  return JSON.stringify(value, null, 2).replace(/[^\x00-\x7F]/g, (char) => {
    const code = char.charCodeAt(0);
    return `\\u${code.toString(16).padStart(4, "0")}`;
  });
}

function cmdLinkGraph(args: LinkGraphArgs): void {
  const notes = iterNotes(args.folder);
  const stemToPath = new Map<string, string>();
  const edges: Array<[string, string]> = [];

  for (const note of notes) {
    stemToPath.set(path.basename(note, path.extname(note)), relPath(note));
  }

  for (const note of notes) {
    const rel = relPath(note);
    let text: string;
    try {
      text = readTextIgnore(note);
    } catch {
      continue;
    }
    for (const match of text.matchAll(/\[\[([^\]|]+)(?:\|[^\]]+)?\]\]/g)) {
      const target = match[1]!.trim();
      const dst = stemToPath.get(target);
      if (dst) edges.push([rel, dst]);
    }
  }

  if (args.format === "dot") {
    console.log("digraph vault {");
    console.log("  rankdir=LR;");
    const seenNodes = new Set<string>();
    for (const [src, dst] of edges) {
      for (const node of [src, dst]) {
        if (!seenNodes.has(node)) {
          const label = path.posix.basename(node, path.posix.extname(node));
          console.log(`  "${node}" [label="${label}"];`);
          seenNodes.add(node);
        }
      }
      console.log(`  "${src}" -> "${dst}";`);
    }
    console.log("}");
  } else {
    const graph = { nodes: [...stemToPath.values()], edges: edges.map(([from, to]) => ({ from, to })) };
    console.log(asciiJson(graph));
  }

  console.error(`\n# ${stemToPath.size} nodes, ${edges.length} edges`);
}

function cmdStructureCheck(args: StructureCheckArgs): void {
  const vault = VAULT_ROOT;
  const issues: string[] = [];
  const expected = args.expected ? new Set(args.expected.split(",")) : new Set<string>();
  const actual = new Set(
    fs.readdirSync(vault, { withFileTypes: true })
      .filter((entry) => entry.isDirectory() && !entry.name.startsWith("."))
      .map((entry) => entry.name),
  );

  if (expected.size) {
    const missing = [...expected].filter((entry) => !actual.has(entry)).sort();
    const extra = [...actual].filter((entry) => !expected.has(entry)).sort();
    if (missing.length) issues.push(`Missing expected folders: ${missing.join(", ")}`);
    if (extra.length) issues.push(`Extra top-level folders: ${extra.join(", ")}`);
  }

  const rootFiles = fs.readdirSync(vault, { withFileTypes: true })
    .filter((entry) => entry.isFile() && !entry.name.startsWith("."))
    .map((entry) => entry.name);
  if (rootFiles.length) issues.push(`Loose files at root: ${rootFiles.join(", ")}`);

  for (const folderName of ["Knowledge", "Projects"]) {
    const folder = path.join(vault, folderName);
    if (fs.existsSync(folder)) {
      let noFm = 0;
      for (const entry of fs.readdirSync(folder, { withFileTypes: true })) {
        if (entry.isFile() && entry.name.endsWith(".md")) {
          const text = readTextIgnore(path.join(folder, entry.name));
          if (!text.startsWith("---")) noFm += 1;
        }
      }
      if (noFm) issues.push(`${folderName}: ${noFm} notes without frontmatter`);
    }
  }

  for (const dir of walkDirs(vault)) {
    if (dir === vault) continue;
    const parts = path.relative(vault, dir).split(path.sep);
    if (parts.some((part) => IGNORE_DIRS.has(part))) continue;
    if (!fs.readdirSync(dir).length) issues.push(`Empty folder: ${path.relative(vault, dir).split(path.sep).join("/")}`);
  }

  for (const obs of walkDirs(vault).filter((dir) => path.basename(dir) === ".obsidian")) {
    if (path.dirname(obs) !== vault) issues.push(`Nested .obsidian config: ${path.relative(vault, obs).split(path.sep).join("/")}`);
  }

  if (issues.length) {
    console.log(`Found ${issues.length} issues:`);
    for (const issue of issues) console.log(`  - ${issue}`);
  } else {
    console.log("Vault structure OK");
  }
}

function walkDirs(root: string): string[] {
  const dirs: string[] = [];
  function visit(dir: string): void {
    dirs.push(dir);
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) visit(path.join(dir, entry.name));
    }
  }
  visit(root);
  return dirs.sort();
}

function cmdFrontmatterFix(args: FrontmatterFixArgs): void {
  const notes = iterNotes(args.folder);
  const issues: Array<[string, string]> = [];
  for (const note of notes) {
    const rel = relPath(note);
    const text = readTextIgnore(note);
    if (!text.startsWith("---")) {
      if (text.trim().startsWith("---")) issues.push([rel, "frontmatter preceded by whitespace"]);
      else if (text.slice(0, 200).includes("---")) issues.push([rel, "frontmatter not on line 1"]);
      else issues.push([rel, "no frontmatter"]);
      continue;
    }

    const end = text.indexOf("\n---", 3);
    if (end === -1) {
      issues.push([rel, "unclosed frontmatter (no closing ---)"]);
      if (!args.dryRun) {
        const lines = text.split("\n");
        let yamlEnd = 1;
        for (let i = 1; i < lines.length; i += 1) {
          const line = lines[i]!;
          if (/^(\w[\w\s-]*:.*|\s+-\s+.*|\s+.*|\s*)$/.test(line)) yamlEnd = i + 1;
          else break;
        }
        lines.splice(yamlEnd, 0, "---");
        fs.writeFileSync(note, lines.join("\n"), "utf8");
        issues[issues.length - 1] = [rel, "unclosed frontmatter — FIXED"];
      }
    }
  }

  if (issues.length) {
    console.log(`Found ${issues.length} frontmatter issues:`);
    for (const [rel, issue] of issues) console.log(`  ${rel}: ${issue}`);
  } else {
    console.log("All frontmatter OK");
  }
}

function parserError(message: string): never {
  console.error(`vault_ops.ts: error: ${message}`);
  process.exit(2);
}

function optionValue(argv: string[], index: number, option: string): string {
  const value = argv[index + 1];
  if (value === undefined) parserError(`argument ${option}: expected one argument`);
  return value;
}

function parseOptions(argv: string[], allowedValueOptions: Set<string>, allowedBooleans = new Set<string>()): { options: Map<string, string | boolean>; positionals: string[] } {
  const options = new Map<string, string | boolean>();
  const positionals: string[] = [];
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i]!;
    if (token.startsWith("--")) {
      if (allowedBooleans.has(token)) {
        options.set(token, true);
      } else if (allowedValueOptions.has(token)) {
        options.set(token, optionValue(argv, i, token));
        i += 1;
      } else {
        parserError(`unrecognized arguments: ${token}`);
      }
    } else {
      positionals.push(token);
    }
  }
  return { options, positionals };
}

function parseArgs(argv: string[]): Args {
  const command = argv[0];
  if (!command) parserError("the following arguments are required: command");
  const rest = argv.slice(1);
  switch (command) {
    case "property-set": {
      const { options, positionals } = parseOptions(rest, new Set(["--folder", "--filter"]));
      if (positionals.length < 2) parserError("the following arguments are required: property, value");
      if (positionals.length > 2) parserError(`unrecognized arguments: ${positionals.slice(2).join(" ")}`);
      return {
        command,
        property: positionals[0]!,
        value: positionals[1]!,
        folder: (options.get("--folder") as string | undefined) ?? null,
        filter: (options.get("--filter") as string | undefined) ?? null,
      };
    }
    case "property-remove": {
      const { options, positionals } = parseOptions(rest, new Set(["--folder"]));
      if (!positionals[0]) parserError("the following arguments are required: property");
      if (positionals.length > 1) parserError(`unrecognized arguments: ${positionals.slice(1).join(" ")}`);
      return { command, property: positionals[0], folder: (options.get("--folder") as string | undefined) ?? null };
    }
    case "property-stats": {
      const { options, positionals } = parseOptions(rest, new Set(["--folder"]));
      if (positionals.length) parserError(`unrecognized arguments: ${positionals.join(" ")}`);
      return { command, folder: (options.get("--folder") as string | undefined) ?? null };
    }
    case "link-graph": {
      const { options, positionals } = parseOptions(rest, new Set(["--folder", "--format"]));
      if (positionals.length) parserError(`unrecognized arguments: ${positionals.join(" ")}`);
      const format = (options.get("--format") ?? "json") as string;
      if (!["dot", "json"].includes(format)) parserError(`argument --format: invalid choice: '${format}' (choose from 'dot', 'json')`);
      return { command, folder: (options.get("--folder") as string | undefined) ?? null, format: format as "dot" | "json" };
    }
    case "structure-check": {
      const { options, positionals } = parseOptions(rest, new Set(["--expected"]));
      if (positionals.length) parserError(`unrecognized arguments: ${positionals.join(" ")}`);
      return { command, expected: (options.get("--expected") as string | undefined) ?? null };
    }
    case "frontmatter-fix": {
      const { options, positionals } = parseOptions(rest, new Set(["--folder"]), new Set(["--dry-run"]));
      if (positionals.length) parserError(`unrecognized arguments: ${positionals.join(" ")}`);
      return { command, folder: (options.get("--folder") as string | undefined) ?? null, dryRun: options.get("--dry-run") === true };
    }
    default:
      parserError(`argument command: invalid choice: '${command}' (choose from 'property-set', 'property-remove', 'property-stats', 'link-graph', 'structure-check', 'frontmatter-fix')`);
  }
}

export function main(): void {
  const args = parseArgs(Bun.argv.slice(2));
  switch (args.command) {
    case "property-set":
      cmdPropertySet(args);
      break;
    case "property-remove":
      cmdPropertyRemove(args);
      break;
    case "property-stats":
      cmdPropertyStats(args);
      break;
    case "link-graph":
      cmdLinkGraph(args);
      break;
    case "structure-check":
      cmdStructureCheck(args);
      break;
    case "frontmatter-fix":
      cmdFrontmatterFix(args);
      break;
  }
}

if (import.meta.main) main();
